include/deal.II/matrix_free/evaluation_kernels.h

```diff
@@ -382,103 +394,102 @@ namespace internal
         temp2 = temp1 + temp_size;
       }
 
+    const unsigned int n_q_points = temp_size == 0 ? shape_info.n_q_points : Eval::n_q_points;
+    const unsigned int dofs_per_comp = (type == MatrixFreeFunctions::truncated_tensor) ?
+                                       Utilities::fixed_power<dim>(shape_info.fe_degree+1) : shape_info.dofs_per_component_on_cell;
     // expand dof_values to tensor product for truncated tensor products
-    VectorizedArray<Number> **values_dofs = values_dofs_actual;
-    VectorizedArray<Number> *expanded_dof_values[n_components];
-    if (type == MatrixFreeFunctions::truncated_tensor)
-      {
-        values_dofs = expanded_dof_values;
-        for (unsigned int c=0; c<n_components; ++c)
-          expanded_dof_values[c] = scratch_data+2*(std::max(shape_info.dofs_per_component_on_cell,
-                                                            shape_info.n_q_points)) +
-                                   c*Utilities::fixed_power<dim>(shape_info.fe_degree+1);
-      }
-
-    // These avoid compiler warnings; they are only used in sensible context but
-    // compilers typically cannot detect when we access something like
-    // gradients_quad[2] only for dim==3.
-    const unsigned int d1 = dim>1?1:0;
-    const unsigned int d2 = dim>2?2:0;
+    Number *values_dofs = (type == MatrixFreeFunctions::truncated_tensor) ?
+                          scratch_data+2*(std::max(shape_info.dofs_per_component_on_cell,
+                                                   shape_info.n_q_points)) :
+                          values_dofs_actual;
 
     switch (dim)
       {
       case 1:
         for (unsigned int c=0; c<n_components; c++)
           {
             if (integrate_values == true)
-              eval.template values<0,false,false> (values_quad[c], values_dofs[c]);
+              {
+                if (overwrite_values == true)
```

Comment:
Changed the name to `add_into_values_array` which better reflects the purpose. We decided not to have a function template within the already templated class for now because it is shorter to write at the call.